File: src/FarmEngine.Core/Farming/Crops.cs
using System.Text.Json;
using FarmEngine.Json;
using FarmEngine.Schemas;

namespace FarmEngine.Core;

/// <summary>
/// Farming math — extracted from src/lib/crops.ts, behavior-identical
/// (characterization-tested), with all randomness going through the seeded
/// RNG instead of Math.random. (Port of engine-core/src/farming/crops.ts.)
/// </summary>
public static class Crops
{
    public static CropDefinition? GetCropDefinitionFromContent(GameContent content, string cropType) =>
        content.Crops.TryGetValue(cropType, out var definition) ? definition : null;

    public static double GetCropStage(double plantedAt, double currentTime, double growthTime, double stages, bool watered = true)
    {
        var elapsed = currentTime - plantedAt;
        var stageTime = growthTime / stages;
        var waterPenalty = watered ? 1 : 0.5;
        var adjustedElapsed = elapsed * waterPenalty;
        return Math.Min(Math.Floor(adjustedElapsed / stageTime), stages - 1);
    }

    public static bool IsCropMature(double stage, double stages) => stage >= stages - 1;

    /// <summary>Returns one of <see cref="CropQualities"/>.</summary>
    public static string CalculateCropQuality(bool watered, bool fertilized, double daysWithoutWater)
    {
        double qualityScore = 0;
        if (watered) qualityScore += 2;
        if (fertilized) qualityScore += 3;
        qualityScore -= daysWithoutWater;

        if (qualityScore >= 4) return CropQualities.Iridium;
        if (qualityScore >= 2) return CropQualities.Gold;
        if (qualityScore >= 1) return CropQualities.Silver;
        return CropQualities.Normal;
    }

    /// <summary>Returns one of <see cref="CropMutations"/> or null. Always draws exactly one float when a definition is given.</summary>
    public static string? RollMutation(CropDefinition? definition, string quality, IRandomSource rng)
    {
        if (definition is null) return null;
        // `definition.mutationChance || 0`
        var baseChance = definition.MutationChance is double chance && chance != 0 && !double.IsNaN(chance) ? chance : 0;
        var qualityBonus = quality == CropQualities.Iridium ? 2 : quality == CropQualities.Gold ? 1.5 : quality == CropQualities.Silver ? 1.2 : 1;
        var finalChance = baseChance * qualityBonus;

        var roll = rng.Float();
        if (roll < finalChance * 0.1) return CropMutations.Ancient;
        if (roll < finalChance * 0.3) return CropMutations.Golden;
        if (roll < finalChance) return CropMutations.Giant;
        return null;
    }

    public static double RollYield(CropDefinition? definition, string quality, string? mutation, IRandomSource rng)
    {
        if (definition is null) return 1;
        var baseYield = rng.Int(definition.YieldMin, definition.YieldMax);
        var qualityBonus = quality == CropQualities.Iridium ? 1.5 : quality == CropQualities.Gold ? 1.3 : quality == CropQualities.Silver ? 1.1 : 1;
        var mutationBonus = mutation == CropMutations.Ancient ? 3 : mutation == CropMutations.Golden ? 2.5 : mutation == CropMutations.Giant ? 2 : 1;
        return Math.Floor(baseYield * qualityBonus * mutationBonus);
    }

    public static double CalculateHarvestValue(CropDefinition? definition, string quality, string? mutation, double quantity = 1)
    {
        if (definition is null) return 0;
        // Unknown keys read `undefined` in JS → NaN through the multiplication.
        var qualityMultiplier = ContentBuiltin.QualityMultipliers.TryGetValue(quality, out var q) ? q : double.NaN;
        var mutationKey = string.IsNullOrEmpty(mutation) ? "none" : mutation;
        var mutationMultiplier = ContentBuiltin.MutationMultipliers.TryGetValue(mutationKey, out var m) ? m : double.NaN;
        return Math.Floor(definition.BaseHarvestValue * qualityMultiplier * mutationMultiplier * quantity);
    }

    public static bool CanGrowInSeason(CropDefinition? definition, string season)
    {
        if (definition is null) return false;
        return definition.Seasons.Contains(season);
    }

    public static string GetCurrentSeason(double gameDay)
    {
        var seasonIndex = Math.Floor(gameDay / ContentBuiltin.DaysPerSeason) % ContentBuiltin.SeasonOrder.Count;
        // JS returns undefined for a negative index; mirror with a bounds check.
        var i = (int)seasonIndex;
        return i >= 0 && i < ContentBuiltin.SeasonOrder.Count ? ContentBuiltin.SeasonOrder[i] : null!;
    }

    public static double GetDayInSeason(double gameDay) => (gameDay % ContentBuiltin.DaysPerSeason) + 1;

    /// <summary>In-game days to maturity for a definition (with legacy-ms fallback).</summary>
    public static double CropGrowthDays(CropDefinition definition)
    {
        if (definition.GrowthDays is double days && days > 0) return days;
        return Math.Min(28, Math.Max(1, Js.Round(definition.GrowthTime / 5000)));
    }

    /// <summary>In-game days between repeat harvests for regrowing crops.</summary>
    public static double CropRegrowthDays(CropDefinition definition)
    {
        if (definition.RegrowthDays is double days && days > 0) return days;
        // `if (definition.regrowthTime)` — 0 and NaN are falsy.
        if (definition.RegrowthTime is double time && time != 0 && !double.IsNaN(time))
            return Math.Min(28, Math.Max(1, Js.Round(time / 5000)));
        return 1;
    }

    /// <summary>Visual growth stage for the day-based model (0 .. stages-1).</summary>
    public static double ComputeCropStage(Crop crop, CropDefinition definition)
    {
        var growthDays = CropGrowthDays(definition);
        var daysGrown = crop.DaysGrown ?? 0;
        var progress = Math.Min(1, daysGrown / growthDays);
        return Math.Min(Math.Floor(progress * (definition.Stages - 1)), definition.Stages - 1);
    }

    /// <summary>Mature when it has accumulated enough watered days.</summary>
    public static bool IsCropMatureByDays(Crop crop, CropDefinition definition)
    {
        if (crop.Withered == true) return false;
        return (crop.DaysGrown ?? 0) >= CropGrowthDays(definition);
    }

    /// <summary>Day-based planting (M2+): crops start unwatered — water them or they won't grow.</summary>
    public static Crop CreatePlantedCrop(string cropType, double plantedOnDay, bool fertilized) => new()
    {
        Type = cropType,
        PlantedAt = 0,
        PlantedOnDay = plantedOnDay,
        DaysGrown = 0,
        Stage = 0,
        Watered = false,
        LastWateredDay = null,
        Quality = fertilized ? CropQualities.Silver : CropQualities.Normal,
        Mutation = null,
        HarvestCount = 0,
        DaysWithoutWater = 0,
    };

    public static Crop InitializeCrop(string cropType, double plantedAt, bool fertilized) => new()
    {
        Type = cropType,
        PlantedAt = plantedAt,
        Stage = 0,
        Watered = true,
        LastWateredDay = 0,
        Quality = fertilized ? CropQualities.Silver : CropQualities.Normal,
        Mutation = null,
        HarvestCount = 0,
        DaysWithoutWater = 0,
    };

    public static bool CanPlaceMultiTileCrop(List<List<Tile>> tiles, double x, double y, double width, double height)
    {
        for (double dy = 0; dy < height; dy++)
        {
            for (double dx = 0; dx < width; dx++)
            {
                var checkX = x + dx;
                var checkY = y + dy;
                if (checkY < 0 || checkY >= tiles.Count || checkX < 0 || checkX >= tiles[0].Count)
                {
                    return false;
                }
                var tile = tiles[(int)checkY][(int)checkX];
                if (tile.Type != TileTypes.Soil || tile.Crop is not null)
                {
                    return false;
                }
            }
        }
        return true;
    }

    /// <summary>Merge built-in crop definitions with a project's custom crops.</summary>
    public static OrderedDictionary<string, CropDefinition> MergeCropDefinitions(IEnumerable<CropDefinition>? customCrops = null)
    {
        var merged = new OrderedDictionary<string, CropDefinition>();
        foreach (var (id, definition) in ContentBuiltin.CropDefinitions) merged[id] = definition;
        foreach (var crop in customCrops ?? [])
        {
            merged[crop.Id] = crop;
        }
        return merged;
    }

    /// <summary>
    /// Overload for <c>project.customCrops</c>: in TS a CustomCropDefinition IS a
    /// CropDefinition (structural typing, passthrough keeps <c>customAsset</c>);
    /// here each one is converted via <see cref="ToCropDefinition"/>.
    /// </summary>
    public static OrderedDictionary<string, CropDefinition> MergeCropDefinitions(IEnumerable<CustomCropDefinition>? customCrops) =>
        MergeCropDefinitions(customCrops?.Select(ToCropDefinition));

    /// <summary>
    /// CustomCropDefinition → CropDefinition via a JSON round-trip, so every key
    /// (including <c>customAsset</c> and passthrough extras) survives in <c>Extra</c>
    /// exactly like the TS object would carry it.
    /// </summary>
    public static CropDefinition ToCropDefinition(CustomCropDefinition custom) =>
        JsonSerializer.Deserialize<CropDefinition>(JsonDefaults.ToElement(custom), JsonDefaults.Options)!;
}

File: crates/farm-sim/src/farming/crops.rs
//! Farming math (port of `Farming/Crops.cs`, engine-core/src/farming/crops.ts). This phase
//! ports the content-assembly part (`mergeCropDefinitions`); the growth, quality and yield
//! functions arrive with the gameplay modules.

use crate::content_builtin;
use crate::schema::{CropDefinition, CustomCropDefinition};
use indexmap::IndexMap;
use serde_json::{Map, Value};

/// Merge built-in crop definitions with a project's custom crops.
pub fn merge_crop_definitions(custom_crops: &[CropDefinition]) -> IndexMap<String, CropDefinition> {
    let mut merged = content_builtin::crop_definitions();
    for crop in custom_crops {
        merged.insert(crop.id.clone(), crop.clone());
    }
    merged
}

/// `project.customCrops` variant: in TS a CustomCropDefinition IS a CropDefinition (structural
/// typing, passthrough keeps `customAsset`); here each one is converted via
/// [`to_crop_definition`].
pub fn merge_custom_crop_definitions(custom_crops: &[CustomCropDefinition]) -> IndexMap<String, CropDefinition> {
    let converted: Vec<CropDefinition> = custom_crops.iter().map(to_crop_definition).collect();
    merge_crop_definitions(&converted)
}

/// CustomCropDefinition → CropDefinition: every key survives, `customAsset` (and any
/// passthrough extras) ride along in `extra` exactly like the TS object would carry them.
pub fn to_crop_definition(custom: &CustomCropDefinition) -> CropDefinition {
    let mut extra = Map::new();
    if let Some(asset) = &custom.custom_asset {
        extra.insert("customAsset".to_owned(), Value::String(asset.clone()));
    }
    for (key, value) in &custom.extra {
        extra.insert(key.clone(), value.clone());
    }
    CropDefinition {
        id: custom.id.clone(),
        name: custom.name.clone(),
        visual: custom.visual.clone(),
        seed_cost: custom.seed_cost,
        base_harvest_value: custom.base_harvest_value,
        growth_time: custom.growth_time,
        growth_days: custom.growth_days,
        stages: custom.stages,
        seasons: custom.seasons.clone(),
        regrowth_time: custom.regrowth_time,
        regrowth_days: custom.regrowth_days,
        can_regrow: custom.can_regrow,
        multi_tile: custom.multi_tile.clone(),
        mutation_chance: custom.mutation_chance,
        yield_min: custom.yield_min,
        yield_max: custom.yield_max,
        extra,
    }
}

/// CropDefinition → CustomCropDefinition (a pack crop stored as a project custom crop): a string
/// `customAsset` in the extras becomes the typed field, like the C# JSON round-trip.
pub fn to_custom_crop_definition(crop: &CropDefinition) -> CustomCropDefinition {
    let mut extra = crop.extra.clone();
    let custom_asset = match extra.get("customAsset") {
        Some(Value::String(asset)) => {
            let asset = asset.clone();
            extra.remove("customAsset");
            Some(asset)
        }
        _ => None,
    };
    CustomCropDefinition {
        id: crop.id.clone(),
        name: crop.name.clone(),
        visual: crop.visual.clone(),
        seed_cost: crop.seed_cost,
        base_harvest_value: crop.base_harvest_value,
        growth_time: crop.growth_time,
        growth_days: crop.growth_days,
        stages: crop.stages,
        seasons: crop.seasons.clone(),
        regrowth_time: crop.regrowth_time,
        regrowth_days: crop.regrowth_days,
        can_regrow: crop.can_regrow,
        multi_tile: crop.multi_tile.clone(),
        mutation_chance: crop.mutation_chance,
        yield_min: crop.yield_min,
        yield_max: crop.yield_max,
        custom_asset,
        extra,
    }
}

// ─── Growth and harvest math (port of the rest of Farming/Crops.cs) ─────────────────────────

use crate::rng::RandomSource;
use crate::schema::{Crop, GameContent, Tile};

pub fn get_crop_definition_from_content<'a>(content: &'a GameContent, crop_type: &str) -> Option<&'a CropDefinition> {
    let _ = (content, crop_type);
    todo!("port Crops.GetCropDefinitionFromContent")
}

pub fn get_crop_stage(planted_at: f64, current_time: f64, growth_time: f64, stages: f64, watered: bool) -> f64 {
    let _ = (planted_at, current_time, growth_time, stages, watered);
    todo!("port Crops.GetCropStage")
}

pub fn is_crop_mature(stage: f64, stages: f64) -> bool {
    let _ = (stage, stages);
    todo!("port Crops.IsCropMature")
}

pub fn calculate_crop_quality(watered: bool, fertilized: bool, days_without_water: f64) -> String {
    let _ = (watered, fertilized, days_without_water);
    todo!("port Crops.CalculateCropQuality")
}

pub fn roll_mutation(definition: Option<&CropDefinition>, quality: &str, rng: &mut dyn RandomSource) -> Option<String> {
    let _ = (definition, quality, rng);
    todo!("port Crops.RollMutation")
}

pub fn roll_yield(
    definition: Option<&CropDefinition>,
    quality: &str,
    mutation: Option<&str>,
    rng: &mut dyn RandomSource,
) -> f64 {
    let _ = (definition, quality, mutation, rng);
    todo!("port Crops.RollYield")
}

pub fn calculate_harvest_value(
    definition: Option<&CropDefinition>,
    quality: &str,
    mutation: Option<&str>,
    quantity: f64,
) -> f64 {
    let _ = (definition, quality, mutation, quantity);
    todo!("port Crops.CalculateHarvestValue")
}

pub fn can_grow_in_season(definition: Option<&CropDefinition>, season: &str) -> bool {
    let _ = (definition, season);
    todo!("port Crops.CanGrowInSeason")
}

pub fn get_current_season(game_day: f64) -> String {
    let _ = game_day;
    todo!("port Crops.GetCurrentSeason")
}

pub fn get_day_in_season(game_day: f64) -> f64 {
    let _ = game_day;
    todo!("port Crops.GetDayInSeason")
}

pub fn crop_growth_days(definition: &CropDefinition) -> f64 {
    let _ = definition;
    todo!("port Crops.CropGrowthDays")
}

pub fn crop_regrowth_days(definition: &CropDefinition) -> f64 {
    let _ = definition;
    todo!("port Crops.CropRegrowthDays")
}

pub fn compute_crop_stage(crop: &Crop, definition: &CropDefinition) -> f64 {
    let _ = (crop, definition);
    todo!("port Crops.ComputeCropStage")
}

pub fn is_crop_mature_by_days(crop: &Crop, definition: &CropDefinition) -> bool {
    let _ = (crop, definition);
    todo!("port Crops.IsCropMatureByDays")
}

pub fn create_planted_crop(crop_type: &str, planted_on_day: f64, fertilized: bool) -> Crop {
    let _ = (crop_type, planted_on_day, fertilized);
    todo!("port Crops.CreatePlantedCrop")
}

pub fn initialize_crop(crop_type: &str, planted_at: f64, fertilized: bool) -> Crop {
    let _ = (crop_type, planted_at, fertilized);
    todo!("port Crops.InitializeCrop")
}

pub fn can_place_multi_tile_crop(tiles: &[Vec<Tile>], x: f64, y: f64, width: f64, height: f64) -> bool {
    let _ = (tiles, x, y, width, height);
    todo!("port Crops.CanPlaceMultiTileCrop")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::stable_json;

    #[test]
    fn custom_crops_override_builtins_by_id_and_keep_custom_asset() {
        let custom = CustomCropDefinition {
            id: "wheat".to_owned(),
            name: "Blue Wheat".to_owned(),
            custom_asset: Some("asset-1".to_owned()),
            ..CustomCropDefinition::default()
        };
        let merged = merge_custom_crop_definitions(std::slice::from_ref(&custom));
        assert_eq!(merged.len(), content_builtin::crop_definitions().len());
        assert_eq!(merged.get_index_of("wheat"), Some(0));
        assert_eq!(merged["wheat"].name, "Blue Wheat");
        assert_eq!(merged["wheat"].extra.get("customAsset"), Some(&Value::String("asset-1".to_owned())));
        // The JSON shapes are identical: the field just moves between typed and extra.
        assert_eq!(stable_json::stringify(&merged["wheat"]), stable_json::stringify(&custom));
        assert_eq!(to_custom_crop_definition(&merged["wheat"]), custom);
    }
}
